### PR TITLE
OAK-8494 - Support AWS Key Managed Service (SSE-KMS)

### PR DESCRIPTION
**Pull request by ankitaagar**
OAK-8494 - Support AWS Key Managed Service (SSE-KMS)

### CODE REVIEW

**Review by amit-jain — COMMENTED** [review]
@ankitaagar 
Without having gone into details some general comments
- Please keep the changes strictly to the feature on hand (Specifically remove any formatting changes)
- I see e.printStackTrace() in the catch blocks, please have proper exception handling

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java [hunk 1]
```diff
@@ -939,10 +941,26 @@ private URI createPresignedURI(DataIdentifier identifier,
             final Date expiration = new Date();
             expiration.setTime(expiration.getTime() + expirySeconds * 1000);
 
-            GeneratePresignedUrlRequest request = new GeneratePresignedUrlRequest(bucket, key)
-                    .withMethod(method)
-                    .withExpiration(expiration);
-
+            GeneratePresignedUrlRequest request = null;
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:944]: This block should also be part of the decorator as all other requests are handled there as well

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java [hunk 2]
```diff
@@ -939,10 +941,26 @@ private URI createPresignedURI(DataIdentifier identifier,
...
+            if (properties.getProperty(S3Constants.S3_ENCRYPTION)
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:945]: `properties.getProperty(S3Constants.S3_ENCRYPTION)` if null will lead to NPE. There should be a check. Also, if not already caught that means relevant unit test missing.

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java [hunk 3]
```diff
@@ -939,10 +941,26 @@ private URI createPresignedURI(DataIdentifier identifier,
...
+                    .equals(S3Constants.S3_ENCRYPTION_SSE_KMS)) {
+                if(properties.getProperty(S3Constants.S3_SSE_KMS_KEYID) != null) {
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Backend.java:947]: Please optimize block to just add the relevant properties for each case as required rather than repeating everything

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Constants.java [hunk 4]
```diff
@@ -102,6 +103,21 @@
...
+
+    /**
+     *  Constant to set keyID for SSE_KMS encryption.
+     */
+    public static final String S3_SSE_KMS_KEYID = "kmsKeyId";
+
+    /**
+     *  Constant to set Signature Type to V4 for AWS KMS
+     */
+    public static final String AWS_S3_V4_SIGNType = "AWSS3V4SignerType";
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3Constants.java:119]: Remove if not required and not used

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java [hunk 5]
```diff
@@ -20,34 +20,64 @@
...
+    Properties props;
+    SSEAwsKeyManagementParams sseParams;
 
     public S3RequestDecorator(Properties props) {
         if (props.getProperty(S3Constants.S3_ENCRYPTION) != null) {
             this.dataEncryption = dataEncryption.valueOf(props.getProperty(S3Constants.S3_ENCRYPTION));
+
+            if (props.getProperty(S3Constants.S3_ENCRYPTION).equals(S3Constants.S3_ENCRYPTION_SSE_KMS)) {
+                String keyId = props.getProperty(S3Constants.S3_SSE_KMS_KEYID);
+                if(!StringUtils.isNullOrEmpty(keyId)){
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:46]: Optimize block to remove redundancies

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java [hunk 6]
```diff
@@ -58,20 +88,63 @@ public PutObjectRequest decorate(PutObjectRequest request) {
...
         switch (getDataEncryption()) {
             case SSE_S3:
-                ObjectMetadata metadata = request.getNewObjectMetadata() == null
+                metadata = request.getNewObjectMetadata() == null
                                 ? new ObjectMetadata()
                                 : request.getNewObjectMetadata();
                 metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
                 request.setNewObjectMetadata(metadata);
                 break;
+            case SSE_KMS:
```
Comment (amit-jain) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:100]: As above optimize block to remove redundancies. Also, better to be consistent i.e. either use switch case or if/else everywhere

File: oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMS.java [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+            expectedNumURIs = 10000;
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+
+            uploadSize = ONE_GB * 200;
+            // expectedNumURIs still 10000, AWS limit
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+        }
+    }
```
Comment (amit-jain) [inline oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMS.java:104]: Missing test case for GeneratePreSignedUrlRequest and possible reason NPE for if condition not detected.

File: oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMSwithKey.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+            expectedNumURIs = 10000;
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+
+            uploadSize = ONE_GB * 200;
+            // expectedNumURIs still 10000, AWS limit
+            upload = ds.initiateDataRecordUpload(uploadSize, -1);
+            Assert.assertEquals(expectedNumURIs, upload.getUploadURIs().size());
+        }
+    }
```
Comment (amit-jain) [inline oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMSwithKey.java:109]: Missing test case for GeneratePreSignedUrlRequest and possible reason NPE for if condition not detected.

File: oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMSwithKey.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+                props.setProperty(S3Constants.S3_ENCRYPTION, S3Constants.S3_ENCRYPTION_SSE_KMS);
+                props.setProperty(S3Constants.S3_RENAME_KEYS, "true");
+                ds = createDataStore();
+
+                rec = ds.getRecord(rec.getIdentifier());
+                Assert.assertEquals(data.length, rec.getLength());
+                assertRecord(data, rec);
+
+                randomGen.nextBytes(data);
+                ds.addRecord(new ByteArrayInputStream(data));
```
Comment (amit-jain) [inline oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMSwithKey.java:81]: assert for record missing before get and after get?

File: oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMS.java [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                props.setProperty(S3Constants.S3_RENAME_KEYS, "true");
+                ds = createDataStore();
+
+                rec = ds.getRecord(rec.getIdentifier());
+                Assert.assertEquals(data.length, rec.getLength());
+                assertRecord(data, rec);
+
+                randomGen.nextBytes(data);
+                ds.addRecord(new ByteArrayInputStream(data));
+                ds.close();
```
Comment (amit-jain) [inline oak-blob-cloud/src/test/java/org/apache/jackrabbit/oak/blob/cloud/s3/TestS3DSWithSSEKMS.java:77]: assert for record missing before get and after get

**Review by amit-jain — CHANGES_REQUESTED** [review]
Haven't run tests but code needs to be better arranged to reduce duplication and increase maintainability.

File: oak-blob-cloud/pom.xml [hunk 11]
```diff
@@ -32,7 +32,7 @@
 
     <properties>
         <netty.version>4.1.17.Final</netty.version>
-        <aws.version>1.11.330</aws.version>
+        <aws.version>1.11.587</aws.version>
```
Comment (mattvryan) [inline oak-blob-cloud/pom.xml:35]: Is this change to the AWS SDK version required for the proposed code change?
I am not opposed to upgrading our AWS SDK dependency version, but I would prefer that this be done in a separate commit if possible.

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java [hunk 12]
```diff
@@ -20,34 +20,64 @@
...
+                metadata = request.getMetadata() == null
                                 ? new ObjectMetadata()
                                 : request.getMetadata();
                 metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
                 request.setMetadata(metadata);
                 break;
+            case SSE_KMS:
+                metadata = request.getMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getMetadata();
```
Comment (mattvryan) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:72]: This code duplicates the same code in the case statement above (for SSE_S3).  It seems like this could be applied to line 60 above, where ObjectMetadata is first declared, since it applies to all cases.  Then just set the algorithm as appropriate in the case statement.

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java [hunk 13]
```diff
@@ -58,20 +88,63 @@ public PutObjectRequest decorate(PutObjectRequest request) {
...
+                metadata = request.getObjectMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getObjectMetadata();
+                metadata.setSSEAlgorithm(ObjectMetadata.AES_256_SERVER_SIDE_ENCRYPTION);
+                request.setObjectMetadata(metadata);
+                break;
+            case SSE_KMS:
+                metadata = request.getObjectMetadata() == null
+                                ? new ObjectMetadata()
+                                : request.getObjectMetadata();
```
Comment (mattvryan) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:130]: Same issue here re: duplicate code in both case statements

File: oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java [hunk 14]
```diff
@@ -58,20 +88,63 @@ public PutObjectRequest decorate(PutObjectRequest request) {
...
+                metadata.setSSEAlgorithm(SSEAlgorithm.KMS.getAlgorithm());
+                request.setObjectMetadata(metadata);
+                String keyId = getSSEParams().getAwsKmsKeyId();
+                request.withSSEAwsKeyManagementParams(keyId == null
+                        ? new SSEAwsKeyManagementParams()
+                        : new SSEAwsKeyManagementParams(keyId));
```
Comment (mattvryan) [inline oak-blob-cloud/src/main/java/org/apache/jackrabbit/oak/blob/cloud/s3/S3RequestDecorator.java:136]: I'd prefer to see a separate method for setting up the appropriate type of metadata in the request, so it is just a simple call from each case, e.g. "request = setRequestEncryptionSSE_KMS(request)" or something along those lines.  This way we isolate this logic in a single place.

**Review by mattvryan — CHANGES_REQUESTED** [review]
